addons/web/static/src/legacy/js/fields/basic_fields.js

```diff
@@ -607,6 +607,13 @@ var FieldChar = InputField.extend(TranslatableFieldMixin, {
         }
         return this._super(value, options);
     },
+
+    isValid: function () {
+        if (this.value && this.field.pattern && !(new RegExp(this.field.pattern)).test(this.value)) {
+            return false;
```

Comment:
Python regex are unicode-aware by default, Javascript's are not, so it's very much possible to have a regex which is accepted by both engines, a value which is correct as far as Python is concerned, and a validation error on the JS side. Which seems less than ideal.

Reply:
I added the `u` flag for the JS check

Reply:
> I added the `u` flag for the JS check

That switches matching to be codepoint-wise (rather than code units) and allow unicode escapes (`\u`) and [unicode property escapes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Unicode_Property_Escapes). It does not make standard character classes (`\w`. `\d`, ...) unicode-aware.

Reply:
> It does not make standard character classes (\w. \d, ...) unicode-aware.

Ouch, this one really sucks.  So even the "common part" of the regexp language does not behave in the same way in both languages?

Reply:
> > It does not make standard character classes (\w. \d, ...) unicode-aware.
> 
> Ouch, this one really sucks. So even the "common part" of the regexp language does not behave in the same way in both languages?

Well it comes down to the fact that JS regex are never unicode-aware -- or at least never required to be (and making them unicode-aware per spec was rejected).

`/u` is misleading because ~~it's javascript~~ it really makes them *codepoint*-aware, in the same way e.g. `String#codePointAt` is the codepoint-aware version of `String#charCodeAt`.

So the Python side should probably use `re.ASCII` to make the normal character classes non-unicode-aware. Though there are still lots of fun divergences e.g. Python has `\N` for named codepoints, and Python requires `\U` for non-BMP non-literal codepoints (Javascript uses a bracketed `\u{}` instead, obviously that only works in `/u` since they're larger than code units).

Reply:
Oh yeah and obviously `/u` mode adds new issues otherwise it wouldn't be fun e.g.
```python
>>> re.compile("\\a")
re.compile('\\a')
```
```js
>> new RegExp('\\a')
/\a/
>> new RegExp('\\a', 'u')
Uncaught SyntaxError: invalid identity escape in regular expression
    <anonymous> debugger eval code:1
```

Reply:
```
re.compile("\\w")
```

Reply:
> Like discussed, I tried this, and it seems to be working correctly [`97e3b104a13e29e59dff114d1c1cc0834f3d5aea..3610a81ea3670e70fb4461e90911d2246031635e` (compare)](https://github.com/odoo/odoo/compare/97e3b104a13e29e59dff114d1c1cc0834f3d5aea..3610a81ea3670e70fb4461e90911d2246031635e)

JS ignores unknown escapes and just returns the following character unmodified so that would just replace "w" by "p{L}" no? The backslashes probably need to be doubled. `String.raw` with a template string is also be an option but for such short strings it seems more overhead than it's worth.

Also probably still need validation on the Python side for escapes we can't / won't support, or conversions if we think them useful (e.g. `\U`).

Reply:
> JS ignores unknown escapes and just returns the following character unmodified so that would just replace "w" by "p{L}" no? The backslashes probably need to be doubled. `String.raw` with a template string is also be an option but for such short strings it seems more overhead than it's worth.

I tried what I implemented of course, and it worked for the strings/regexes I tested (including unicode)

Reply:
> I tried what I implemented of course, and it worked for the strings/regexes I tested (including unicode)

Because what you implemented replaces e.g. the string `"w"` by the string `"p{L}"`:
```js
>> "water".replace('\w', '\p{L}')
              .replace('\W', '\P{L}')
              .replace('\d', '\p{N}')
              .replace('\D', '\P{N}');
"p{L}ater"
```
An other issue with the implementation is `String#replace(String, String)` only replaces the first occurrence / match:
```js
>> "wwww".replace('\w', '\p{L}');
"p{L}www"
```
you need to use the regex form with the `/g` flag to perform global replacements.

Reply:
Good catch, thanks! Should be good now then.